sdks/java/extensions/sql/datacatalog/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datacatalog/DataCatalogClientAdapter.java

```diff
@@ -31,63 +30,51 @@
 import io.grpc.ManagedChannelBuilder;
 import io.grpc.MethodDescriptor;
 import io.grpc.auth.MoreCallCredentials;
-import java.io.IOException;
+import java.util.function.Supplier;
 import javax.annotation.Nullable;
+import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
 import org.apache.beam.sdk.extensions.sql.meta.Table;
 
 /** Wraps DataCatalog GRPC client and exposes simplified APIS for Data Catalog Table Provider. */
 class DataCatalogClientAdapter {
 
-  private DataCatalogBlockingStub dcClient;
+  private DataCatalogPipelineOptions options;
 
-  private DataCatalogClientAdapter(DataCatalogBlockingStub dcClient) {
-    this.dcClient = dcClient;
+  private Supplier<DataCatalogBlockingStub> dcClient = () -> newClient(options);
+
+  private DataCatalogClientAdapter(DataCatalogPipelineOptions options) {
+    this.options = options;
   }
 
   /** Prod endpoint (default set in pipeline options): datacatalog.googleapis.com. */
-  public static DataCatalogClientAdapter withDefaultCredentials(String endpoint)
-      throws IOException {
-    return new DataCatalogClientAdapter(newClient(endpoint));
+  public static DataCatalogClientAdapter withOptions(DataCatalogPipelineOptions options) {
+    return new DataCatalogClientAdapter(options);
   }
 
-  private static DataCatalogBlockingStub newClient(String endpoint) throws IOException {
+  private DataCatalogBlockingStub newClient(DataCatalogPipelineOptions options) {
```

Comment:
Client setup isn't free, maintaining an existing channel that has already been through auth and possibly has an active http connection will be noticeable from a performance perspective and load on datacatalogs service.

Swap the supplier to be wrapped with Guava's memoizing supplier or JDK equivalent.

Reply:
Yea, good call. At first I didn't wrap it because it was a ThrowingSupplier where we don't have the nice utilities already. But turns out it didn't need to be, so that's easy.

Reply:
TBH it seems like a "client adapter" should have a separate client per adapter. But I'm not really convinced that this class needs to exist at all...

Reply:
Actually, I think the client and the adapter should have the same lifetime if the client is an internal detail. And each new `DataCatalogClientAdapter` could have different pipeline options, so you would need to manage multiple underlying clients.

I tried the alternative of injecting a client, so the caller could own the configuration and lifecycle. And that made this class pretty pointless, so I just inlined the whole class and some other trivial methods.